server/src/quotas.rs

```diff
@@ -0,0 +1,373 @@
+use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};
+
+use failure::Fail;
+use r2d2::Pool;
+
+use crate::actors::project::{Quota, RedisQuota, RejectAllQuota, RetryAfter};
+
+/// The ``grace`` period allows accomodating for clock drift in TTL
+/// calculation since the clock on the Redis instance used to store quota
+/// metrics may not be in sync with the computer running this code.
+static GRACE: u64 = 60;
+
+#[derive(Debug, Clone, Copy)]
+struct UnixTimestamp(u64);
+
+#[derive(Clone)]
+pub enum RedisPool {
+    Cluster(Pool<redis::cluster::ClusterClient>),
```

Comment:
Unhappy about this and how it's used. Couldn't generalize properly.

Reply:
This is totally fair, imho. The only way I can see here is to hide this and expose a higher-level `Quotas` service object that internally deals with Redis and all quirks around where and when to apply which quotas. 

In a sense, this is already done with `is_rate_limited`, except that one has to construct and pass in the redis client.